File: lib/document-parser.ts
import fs from "fs";
import path from "path";

export function chunkText(text: string, chunkSize = 700, overlap = 120): string[] {
  const words = text.split(/\s+/).filter(Boolean);
  if (!words.length) return [];
  const chunks: string[] = [];
  let start = 0;
  while (start < words.length) {
    const end = Math.min(start + chunkSize, words.length);
    const chunk = words.slice(start, end).join(" ").trim();
    if (chunk) chunks.push(chunk);
    if (end >= words.length) break;
    start = Math.max(0, end - overlap);
  }
  return chunks;
}

export function topKeywords(text: string, maxKeywords = 5): string[] {
  const stop = new Set(["the", "a", "an", "is", "of", "for", "and", "to", "in", "on", "with", "that", "it", "this", "as"]);
  const freq: Record<string, number> = {};
  for (const raw of text.toLowerCase().split(/\s+/)) {
    const token = raw.replace(/\W/g, "");
    if (token.length < 4 || stop.has(token)) continue;
    freq[token] = (freq[token] ?? 0) + 1;
  }
  return Object.entries(freq)
    .sort((a, b) => b[1] - a[1])
    .slice(0, maxKeywords)
    .map(([t]) => t);
}

export function estimateDifficulty(wordCount: number): string {
  if (wordCount < 500) return "easy";
  if (wordCount < 2000) return "medium";
  return "hard";
}

export async function parseDocument(filePath: string): Promise<string> {
  const ext = path.extname(filePath).toLowerCase();
  const buf = fs.readFileSync(filePath);

  if ([".txt", ".md"].includes(ext)) {
    return buf.toString("utf8").replace(/\r\n/g, "\n");
  }

  if (ext === ".pdf") {
    const pdfParse = (await import("pdf-parse")).default;
    const data = await pdfParse(buf);
    return (data as { text: string }).text?.trim() ?? "";
  }

  if (ext === ".docx") {
    const mammoth = await import("mammoth");
    const result = await mammoth.extractRawText({ buffer: buf });
    return result.value?.trim() ?? "";
  }

  throw new Error(`Unsupported file type: ${ext}`);
}

File: backend/app/services/document_parser.py
"""Document parsing and chunking (aligned with lib/document-parser.ts)."""
from pathlib import Path


def chunk_text(text: str, chunk_size: int = 700, overlap: int = 120) -> list[str]:
    words = [w for w in text.split() if w]
    if not words:
        return []
    chunks: list[str] = []
    start = 0
    while start < len(words):
        end = min(start + chunk_size, len(words))
        chunk = " ".join(words[start:end]).strip()
        if chunk:
            chunks.append(chunk)
        if end >= len(words):
            break
        start = max(0, end - overlap)
    return chunks


def parse_document(file_path: str | Path) -> str:
    path = Path(file_path)
    ext = path.suffix.lower()
    raw = path.read_bytes()

    if ext in (".txt", ".md"):
        return raw.decode("utf-8", errors="replace").replace("\r\n", "\n")

    if ext == ".pdf":
        from pypdf import PdfReader
        import io
        reader = PdfReader(io.BytesIO(raw))
        parts = []
        for page in reader.pages:
            t = page.extract_text()
            if t:
                parts.append(t)
        return "\n".join(parts).strip()

    if ext == ".docx":
        from docx import Document
        import io
        doc = Document(io.BytesIO(raw))
        return "\n".join(p.text for p in doc.paragraphs if p.text.strip()).strip()

    raise ValueError(f"Unsupported file type: {ext}")
